### DISCUSSION TITLE
Simplify internals

### DISCUSSION THREAD

**Discussion by esatterwhite** [esatterwhite/node-seeli#32] [Ideas]
Seeli is the first node package I wrote and it has accumulated a bit of cruft over the years. The way I want to use it has change over the years and some of the initial decisions on implementation are getting rather difficult to work around

* expectations around command instances are difficult to test
* The sub command interface is difficult to juggle and interferes with trying to programmatically execute a command
* The explicit use of `console.log` make it difficult to test + stub a writable destination
* The config interface can cause some when loading order isn't respected
  * The dynamic loading of the help command may get the wrong color / name if loaded too soon
* The attempt at deep merging of options is probably unnecessary. Most of the options are flat with the exception of flags. I most try to work around it. shallow is probably fine
*  

If the command runner loaded classes rather than instances, many problems would be mitigated.
This would probably be easier with class properties

**Comment by esatterwhite** [comment]
```javascript
const {Command} = require('seeli')

class MyCommand extends Command {
  name = 'mycmd'
  description = 'hello world'
  string = false
  usage = [
    String
  , String
  ]
  commands = [
    Command
  , Command
  ]

  async run(data){
    
  }
}
```

This would be a bit easier to work with, but defining sane defaults becomes a bit more difficult. It would really require the end user to know quite a bit about the internals and have to manually re-use them.

**Reply by esatterwhite** [reply]
This is more the oclif model. But I don't like that it is tied to file structure. I prefer to load commands + sub commands

**Reply by esatterwhite** [reply]
There should be an internal `log` function that writes to the configured stream.
Currently, you can return anything from the `run` function and delegate to a content handler. 

I actually like that you can do this, its great for testing to be able to return objects. however, the content being rendered during the run function is usually handled by `console.log` or `ora` neither of which you can easily stub.

Being able to configure a stream would allow you to more easily silence the commands for tests and still get out.

**Reply by evanlucas** [reply]
I really like this idea.

**Comment by esatterwhite** [comment]
Configuration being a global thing is kind of difficult. I think this would be better initialized from `pakcage.json`

```json5
// package.json
{
  "name": 'a-cli'
, "version": "0.0.1"
, "seeli": {
    "color": "blue"
  , "name": "whizbang"
  , "exitOnContent": false
  , "stream": 1
  }
}
```

This would make it easy to initialize defaults at start up instead of forcing the end user to do it

**Reply by esatterwhite** [reply]
The defaults should come from the global config / package.json, But I would want to be able to pass overrides manually when constructing for testing

```javascript
const MyCommand = require('./my-command')

const cmd = new MyCommand({
  description: 'alternate description'
, stream: new stream.PassThrough()
})

await cmd.run({...data})
```

**Reply by esatterwhite** [reply]
config addressed in #34

**Comment by esatterwhite** [comment]
replace inquirer with [enquirer](https://www.npmjs.com/package/enquirer)

This would bring the number of dependencies down quite a bit, it only has 1. inquirer has 15
